src/generators/notebook/set_collapsed_state.ts

```diff
@@ -0,0 +1,157 @@
+      }
+      // Collapse/expand a sub-cell by setting the its `hidden` state:
+      w.setHidden(state);
+      continue;
+    }
+    // Code cells are more involved, as we need to analyze the outputs to check for generated Markdown/HTML...
+    const c = w as CodeCell;
+    const model = w.model as CodeCellModel;
+    const outputs = model.outputs;
+
+    // First, we do an initial pass to check for generated Markdown/HTML. If we don't find Markdown/HTML, then the entire cell can be collapsed/expanded (both inputs and outputs)...
+    let FLG = false;
+    let dtypes: string[] = [];
+    for (let j = 0; j < outputs.length; j++) {
+      // Retrieve the cell output model:
+      const m = outputs.get(j);
+
+      // Check the cell output MIME types:
+      dtypes = Object.keys(m.data);
+      for (let k = 0; k < dtypes.length; k++) {
+        const t = dtypes[k];
+        if (isMarkdown(t) || isDOM(t) || t === 'text/plain') {
+          // FIXME: apparent Jupyter bug where additional Markdown displays have a `text/plain` MIME type
+          FLG = true;
+        } else {
+          dtypes[k] = '';
+        }
+      }
+    }
+    // If we did not find Markdown/HTML, collapse/expand the entire cell...
+    if (FLG === false) {
+      w.setHidden(state);
+      continue;
+    }
+    // Now, we perform a second pass to determine whether the output areas containing generated Markdown/HTML contain headings at the same or higher level...
+    let idx = -1;
+    for (let j = 0; j < outputs.length; j++) {
+      if (dtypes[j] === '') {
+        continue;
+      }
+      // Retrieve the output area widget:
+      const ow = c.outputArea.widgets[j] as Panel;
+
+      // Determine the heading level from the rendered HTML of the output area result:
+      const h = parseHeading(ow.widgets[1].node.innerHTML);
+
+      // Check if an output widget contains a heading at the same or higher level...
+      if (h && h.level >= 0 && h.level <= level) {
+        // We've reached the end of the section...
+        idx = j;
+        break;
+      }
+    }
+    // If we did not encounter a new section, we can safely collapse/expand the entire widget...
+    if (idx === -1) {
+      w.setHidden(state);
+      continue;
+    }
+    // Finally, we perform a third pass to collapse/expand individual output area widgets...
+    for (let j = 0; j < idx; j++) {
+      const ow = c.outputArea.widgets[j] as Panel;
+      ow.setHidden(state);
+    }
+    // Collapse/expand a sub-cell's input area by setting the its `hidden` state (NOTE: a collapsed input area will still appear collapsed!!! Once a user clicks on the collapsed cell ellipses will the input area display (or, rather, not display!) as being fully hidden. If we want an input area to always be present, e.g., to always allow a user the ability to edit an expanded section, whether or not generated sections are collapsed or expanded, we could remove the following line.):
+    // w.inputArea.setHidden(state);
```

Comment:
Is this intended to be commented out?

Reply:
Yes, I added a comment. In short, was not absolutely clear to me the desired behavior. Ended up deciding to avoid collapsing/expanding a sub-cell's input area because I preferred the rendered output better. However, I wanted to leave in the code comments so that a future dev reading the code might more easily discover the option that rendering could be different or even preferred.